core/templates/dev/head/exploration_editor/EditorServices.js

```diff
@@ -1952,3 +2012,283 @@ oppia.factory('explorationWarningsService', [
     };
   }
 ]);
+
+oppia.factory('lostChangesService', ['utilsService', function(utilsService) {
+  var CMD_ADD_STATE = 'add_state';
+  var CMD_RENAME_STATE = 'rename_state';
+  var CMD_DELETE_STATE = 'delete_state';
+  var CMD_EDIT_STATE_PROPERTY = 'edit_state_property';
+
+  var getRulesList = function(value) {
+    var rulesList = [];
+    value.rule_specs.forEach(function(ruleSpec) {
+      var ruleElm = angular.element('<div></div>');
+      ruleElm.append('<p>Type: ' + ruleSpec.rule_type + '</p>');
+      ruleElm.append(
+        '<p>Value: ' +
+          (
+            Object.keys(ruleSpec.inputs).map(function(input) {
+              return ruleSpec.inputs[input];
+            })
+          ).toString() +
+          '</p>');
+      rulesList.push(ruleElm);
+    });
+
+    return rulesList;
+  };
+
+  var getStatePropertyValue = function(
```

Comment:
Is there some weird feature of the way the edits are represented that means you need this? If so then a comment would be helpful.